### PR TITLE
Resolve #40 and demo for VL20

### LINKED ISSUE #40: UncertainQtty unexpected result
### Bug Description:

Motivated by the [output](https://github.com/nansencenter/DAPPER/blob/1b63a28e07644223a0e68147a5ae6a17be2e37dc/tests/test_example_2.py#L63) in `test_example_2.py`, using 7.7676 ±1.2464 as input:
```python
>>> from dapper.tools.rounding import UncertainQtty
>>> uq = UncertainQtty(7.7676, 1.2464)
>>> print(uq)
UncertainQtty(val=8, prec=1.0)
```
This is in contrast to the expected `UncertainQtty(val=8.0, prec=1.0)` or `UncertainQtty(val=8, prec=1)`. I think the latter is preferred.

### Cause of the bug:
Class `UncertainQtty` prints the first significant figure of the `prec` with the following code:
```python
>>> np.round(1.2464 ,0)
1.0
>>> np.round(0.2464 ,0)
0.0
```

### Possible Solutions:
Turn `float` to `int` if the first significant figure is ahead of the decimal separator.

### Note
This issue is related to an old version TODO, which speculates the use of `uq` in the `xps.tabulate_avrgs` and `xpSpace.print`. 

The difficulty is that, `uq` cannot be used directly for the output because it does not allow the designated `decimals` for its output. [`test_example_2.py`](https://github.com/nansencenter/DAPPER/blob/d53610b476ebf0cd68b2085dc072cac0e0cb00cd/tests/test_example_2.py#L26) has examples for `decimals = 4`.

Any comments?

### PR DESCRIPTION
**Pull request by yumengch**
## Aim

Add bib and demo for VL20

Resolve issue #40 

## Potential issue

- Clean-up for function [`preprocess`](https://github.com/yumengch/DAPPER/blob/ab9ef8358042951b9eb150d58ce96dc7da4ba0d7/dapper/stats.py#L525) in `align_col` needed? 
- 'frmt' in 'align_col' must be given carefully, not an issue for now.
- [This](https://github.com/yumengch/DAPPER/blob/ab9ef8358042951b9eb150d58ce96dc7da4ba0d7/tests/test_data.py#L1232) used to be `0.03` now `0.030` based on `prec` though `prec` is not shown. I feel this is correct.

### CODE REVIEW

File: dapper/mods/VL20/__init__.py [hunk 1]
```diff
@@ -1,10 +1,12 @@
 """
-One scale modified model of Lorenz96 which added one thermodynamic variable.
+One scale modified model of Lorenz96 with an added thermodynamic variable.
```
Comment (patricknraanes) [inline dapper/mods/VL20/__init__.py:2]:
```suggestion
Single-scale Lorenz-96 with an added thermodynamic component.
```
Would that be correct?
Also, no newline at top.

Comment (yumengch) [reply dapper/mods/VL20/__init__.py:2]: It is correct and better. Thanks for the suggestion.

File: dapper/mods/VL20/__init__.py [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
-One scale modified model of Lorenz96 which added one thermodynamic variable.
+Single-scale Lorenz-96 with an added thermodynamic component.
+
+Refs: `bib.vissio2020mechanics`
 """
 import numpy as np
 
 
-class model():
+class model_instance():
```
Comment (patricknraanes) [inline dapper/mods/VL20/__init__.py:9]: This is probably not the best class name. I don't know why I didn't follow convention. I suggest renaming to ModelConfig

This should also be done to QG and LUV models, but maybe in another PR

Comment (yumengch) [reply dapper/mods/VL20/__init__.py:9]: I agree ModelConfig seems better but I think this is something we should take a note for the next PR related to parameter estimation and documentation migration, which require some thoughts on the changes of mods.

File: dapper/stats.py [hunk 3]
```diff
@@ -518,7 +518,8 @@ def align_col(col, header, pad='␣', missingval='', frmt=None):
     - Use tabulate() to get decimal point alignment.
     - Inf and nan are handled individually so that they don't
       align left of the decimal point (makes too wide columns).
-      Custom ``frmt`` also supported.
+      Custom ``frmt`` also supported, which must support formatting
```
Comment (patricknraanes) [inline dapper/stats.py:521]: Single tick?

File: dapper/tools/rounding.py [hunk 4]
```diff
@@ -74,7 +74,8 @@ def round(self, mult=1.0):  # noqa
     def __str__(self):
         """Returns 'val ±prec'.
 
-        The value string contains the number of decimals required by the confidence.
+        The value string contains the number of decimals required
+        by the confidence.
```
Comment (patricknraanes) [inline dapper/tools/rounding.py:78]:
```suggestion
        The value is formatted with the appropriate number of decimals, considering the precision.
```

File: dapper/tools/rounding.py [hunk 5]
```diff
@@ -83,8 +84,10 @@ def __str__(self):
             # Rounding to fallback (rc.sigfig) already took place
             return f"{v} ±{c}"
         if c == 0:
-            # 0 (i.e. not 1e-300) never arises "naturally" => Treat it "magically"
-            # by truncating to a default. Also see https://stackoverflow.com/a/25899600
+            # 0 (i.e. not 1e-300) never arises "naturally"
+            # => Treat it "magically"
+            # by truncating to a default.
+            # Also see https://stackoverflow.com/a/25899600
```
Comment (patricknraanes) [inline dapper/tools/rounding.py:90]: There is no need for these line breaks. The max_line_length is set to 88 is pyproject.toml

File: dapper/tools/rounding.py [hunk 6]
```diff
@@ -93,6 +96,10 @@ def __str__(self):
         if n < 0:
             # Ensure we get 1.30 ±0.01, NOT 1.3 ±0.01:
             frmt = "%%0.%df" % -n
+
+        if n >= 0 and not np.isnan(c) and not np.isinf(c):
```
Comment (patricknraanes) [inline dapper/tools/rounding.py:100]:
```suggestion
        elif np.isfinite(c):
```
(the nan check is already taken care of on line 82, but we do need to check for infinity)

File: dapper/tools/rounding.py (hunk 5 above)
Comment (yumengch) [reply dapper/tools/rounding.py:90]: Thanks, we can restore this.

File: dapper/stats.py [hunk 7]
```diff
@@ -577,20 +596,36 @@ def unpack_uqs(uq_list, decimals=None, cols=("val", "prec")):
     decimals: int
         Desired number of decimals.
         Used for the columns "val" and "prec", and only these.
-        If `None`, it is left to the `dapper.tools.rounding.UncertainQtty` objects.
+        If `None`, it is left to the
+        `dapper.tools.rounding.UncertainQtty` objects.
```
Comment (patricknraanes) [inline dapper/stats.py:600]: No linebreak

File: dapper/tools/rounding.py [hunk 8]
```diff
@@ -83,8 +83,10 @@ def __str__(self):
             # Rounding to fallback (rc.sigfig) already took place
             return f"{v} ±{c}"
         if c == 0:
-            # 0 (i.e. not 1e-300) never arises "naturally" => Treat it "magically"
-            # by truncating to a default. Also see https://stackoverflow.com/a/25899600
+            # 0 (i.e. not 1e-300) never arises "naturally"
+            # => Treat it "magically"
+            # by truncating to a default.
+            # Also see https://stackoverflow.com/a/25899600
```
Comment (patricknraanes) [inline dapper/tools/rounding.py:89]:
```suggestion
            # 0 (i.e. not 1e-300) never arises "naturally" => Treat it "magically"
            # by truncating to a default. Also see https://stackoverflow.com/a/25899600
```

File: dapper/tools/rounding.py [hunk 9]
```diff
@@ -93,6 +93,10 @@ def __str__(self):
         if n < 0:
             # Ensure we get 1.30 ±0.01, NOT 1.3 ±0.01:
             frmt = "%%0.%df" % -n
+
+        elif np.isfinite(c):
+            # turn float point value to integer for |c| >= 1.0
```
Comment (patricknraanes) [inline dapper/tools/rounding.py:98]:
```suggestion
            # if c >= 1.0
```

File: dapper/stats.py (hunk 7 above)
Comment (patricknraanes) [inline dapper/stats.py:600]:
```suggestion
        If `None`, it is left to the `dapper.tools.rounding.UncertainQtty` objects.
```

**Review by patricknraanes — APPROVED** [review]
Great work!

The result (the changes in `test_data.py`) of your improvements is exactly as desired. However, I feel like the combo `unpack_uqs` and `align_col` are too complicated. I will have a look at it after this, to see if I can simplify them.

If you feel it's ready, you can add `VL20` to the list of models in the README. And maybe add your institute's logo as well, in the contributors section.

Also, the `d2x_dtdx` should be tested in `tests_TLMs.py`

**Comment by patricknraanes** [comment]
> I feel like the combo `unpack_uqs` and `align_col` are too complicated. I will have a look at it after this, to see if I can simplify them.

Here are my changes to improve tabulation of results.
https://github.com/nansencenter/DAPPER/compare/cd47cd0..e9c5535
Note that there are further improvements in the output of `tests/test_data.py`.

The main implementation changes are
- `align_col` no longer relies on tabulate. Instead it does the decimal-point alignment itself. This avoids roundabout pre/post-processing.
- All special formatting has been moved from `align_col` to `unpack_uqs`.
- _All_ of the attrs of `uq`s are extracted by `unpack_uqs`.
- The above allowed to greatly simplify `xp_process.py:xpSpace.print:make_cols`

**Comment by yumengch** [comment]
Great work, Patrick!

Some comments:
- Assigning [`str` type values to `uq.val` and `uq.prec`](https://github.com/nansencenter/DAPPER/blob/dda3501c6d269adf3afa3d060520b13afc3e8ddc/dapper/stats.py#L609) leads to error in `uq.__str__`. I'm not sure if it is necessary to safeguard `uq.__str__` for the `val` and `prec` types. If this problem is limited within `unpack_uq`, I think we can simply ignore this.
- I did some tests for `unpack_uqs`:
```python
>>> import numpy as np
>>> from dapper.tools.rounding import UncertainQtty as UQ
>>> from dapper.stats import unpack_uqs, align_col
>>> vals = np.array([0.43085084, 0.9192679 , 0.87700933, 2., 0.1])
>>> precs = np.array([0.08460475, 0.76110353, 0.001, 1.45934142, 1.35481388])
>>> uq_list = [UQ(val, prec) for val, prec in zip(vals, precs)]
>>> unpack_uqs(uq_list, decimals=None)
{'val': ['0.43', '0.9', '0.877', '2', '0'], 'prec': ['0.08', '0.8', '0.001', '1', '1']}
>>> unpack_uqs(uq_list, decimals=4)
{'val': [0.4309, 0.9193, 0.877, 2.0, 0.1], 'prec': [0.0846, 0.7611, 0.001, 1.4593, 1.3548]}
>>> col = unpack_uqs(np.array([None, None, None]), decimals=None)
>>> align_col(['rand']+col['val'])
['rand', '␣␣␣␣', '␣␣␣␣', '␣␣␣␣']
>>> align_col(['1s']+col['prec'])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'prec'
```
It is worth noting two special cases:
-- `0.87700933 ± 0.001` and `2.0 ± 1.0` don't generate `0.8770±0.0010` and `2.0000±1.0000` when I specify `decimals=4`. This is again an issue of necessity to padding `0`s for truncated values.
-- Regarding your [`TODO 3: validate`](https://github.com/nansencenter/DAPPER/blob/dda3501c6d269adf3afa3d060520b13afc3e8ddc/dapper/stats.py#L621),  the combination of `unpack_uq` and `align_col` in the `xp_process` should be fine, but may have error in [`tabulate_avrgs`](https://github.com/nansencenter/DAPPER/blob/dda3501c6d269adf3afa3d060520b13afc3e8ddc/dapper/stats.py#L644), but I'm not so sure if `tabulate_avrgs` needs to deal with such input.

**Comment by patricknraanes** [comment]
Many thanks for the comments.

> If this problem is limited within `unpack_uq`, I think we can simply ignore this.

It is contained within, yes, thanks to `deepcopy`. However, I changed this to `vars(uq).copy()` [the most recent commit](https://github.com/nansencenter/DAPPER/commit/cff8032), which is maybe more explicit/safe solution.


> don't generate `0.8770±0.0010` and` 2.0000±1.0000` when I specify `decimals=4`

Fixed in the latest commit. This fix was simple now that we don't have to worry about `tabulate` truncating zeros.

> Regarding your TODO 3: validate

This is now gone, I believe. Not sure if a similar issue could exist. But it seems good enough for now.

**Comment by yumengch** [comment]
It looks great now. Thanks.

**Pull request merged**